### PR TITLE
Added Branch And Bound algorithm to solve 0-1 Knapsack Problem

### PR DESCRIPTION
**Pull request by yuanjie8629**
### **Describe your change:**
Added Branch And Bound Algorithm to solve 0-1 Knapsack Problem

* [x] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [x] #4774

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by nivedha02 — APPROVED** [review]
well done

**Review by nivedha02 — CHANGES_REQUESTED** [review]
Great work

**Comment by poyea** [comment]
@algorithms-keeper review-all

File: knapsack/branch_and_bound/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+The least cost(LC) is considered the most intelligent as it selects the next node based on a Heuristic Cost Function. It picks the one with the least cost.
+As 0/1 Knapsack is about maximizing the total value, we cannot directly use the LC Branch and Bound technique to solve this. Instead, we convert this into a minimization problem by taking negative of the given values.
+Follow the steps below to solve the problem:
+
+1. Sort the items based on their value/weight(V/W) ratio.
```
Comment (poyea) [inline knapsack/branch_and_bound/README.md:17]: It'd be helpful if we make it \`V\` and \`W\`. Same for all other variables like minLB -> \`minLB\` -> `minLB`

File: knapsack/branch_and_bound/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+The data is seperated by ','
+First data is the name of item.
+Second data is the weight of item.
+Third data is the rating of the item.
+(Name,Weight,Rating)
+
+## Tests
+
+To be completed.
```
Comment (poyea) [inline knapsack/branch_and_bound/README.md:57]: I think you've completed?

File: knapsack/branch_and_bound/branch_and_bound.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+There are total of 6 items as follows:
+1. Ant Repellent,1,2
+2. Blanket,4,3
+3. Brownies,3,10
+4. Frisbee,1,6
+5. Salad,5,4
+6. Watermelon,10,10
+
+Calculate the maximum rating and total weight that the knapsack can take.
+"""
```
Comment (poyea) [inline knapsack/branch_and_bound/branch_and_bound.py:22]: Or we can put these in README, making the code itself "self-contained"

File: knapsack/branch_and_bound/item.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+class Item:
+    def __init__(self, name: str, weight: float, rating: float) -> None:
+        self._name = name
+        self._weight = weight
+        self._rating = rating
```
Comment (poyea) [inline knapsack/branch_and_bound/item.py:5]: `rating`means `value` I guess? See if we can align it with the definition in README.

**Comment by cclauss** [comment]
In preparation for that denial-of-service attack that is [Hacktoberfest](https://hacktoberfest.com), I am closing all pull requests that do not have passing tests.

**Pull request closed without merging**